src/repl-frontend/scala/tools/nsc/interpreter/shell/Reporter.scala

```diff
@@ -0,0 +1,225 @@
+
+  import scala.io.AnsiColor.{ RED, YELLOW, RESET }
+
+  private def label(severity: Severity): String = severity match {
+    case ERROR   => "error"
+    case WARNING => "warning"
+    case INFO    => ""
+  }
+
+  protected def clabel(severity: Severity): String = label(severity) match {
+    case "" => ""
+    case s  => s"$s: "
+  }
+
+  def severityColor(severity: Severity): String = severity match {
+    case ERROR   => RED
+    case WARNING => YELLOW
+    case INFO    => RESET
+  }
+
+  def print(pos: Position, msg: String, severity: Severity): Unit = {
+    val prefix =
+      if (colorOk) severityColor(severity) + clabel(severity) + RESET
+      else clabel(severity)
+
+    printMessage(pos, prefix + msg)
+  }
+
+  private val indentation = " " * indentDepth
+  private def indented(str: String) = str.lines.mkString(indentation, "\n" + indentation, "")
+
+  // indent errors, error message uses the caret to point at the line already on the screen instead of repeating it
+  // TODO: can we splice the error into the code the user typed when multiple lines were entered?
+  // (should also comment out the error to keep multi-line copy/pastable)
+  // TODO: multiple errors are not very intuitive (should the second error for same line repeat the line?)
+  // TODO: the console could be empty due to external changes (also, :reset? -- see unfortunate example in jvm/interpeter (plusOne))
+  def printMessage(posIn: Position, msg: String): Unit = {
+    if ((posIn eq null) || (posIn.source eq NoSourceFile)) printMessage(msg)
+    else {
+      val consoleLinePrefix = "On line "
+      val locationPrefix =
+        posIn.source.file.name match { case  "<console>" => consoleLinePrefix case n => s"$n:" }
+
+      // If there's only one line of input, and it's already printed on the console (as indicated by the position's source file name),
+      // reuse that line in our error output, and suppress the line number (since we know it's `1`)
+      // TODO: this is not perfect, see e.g. test/files/run/repl-colon-type.scala,
+      // where the error refers to a line that's not on the screen
+      if (locationPrefix == consoleLinePrefix &&
+        lastInputSingleLine &&
+        posIn.line > preambleLineDelta) { // line won't be printed if it was part of the preamble
+        printMessage(indentation + posIn.lineCaret)
+        printMessage(indented(msg))
+      }
+      else {
+        val (msgFirstLine, msgRest) =
+          msg.indexOf('\n') match {
+            case -1 => (msg, "")
+            case n => (msg.substring(0, n), msg.substring((n + 1) min msg.length))
+          }
+
+        // add newline to get away from prompt when we're reporting on a script/paste
+        if (locationPrefix != consoleLinePrefix) printMessage("")
+
+        printMessage(indentation + posIn.lineContent)
+        printMessage(indentation + posIn.lineCaret)
+        printMessage(s"$locationPrefix${posIn.line - preambleLineDelta}: $msgFirstLine")
```

Comment:
The error message you're seeing does look wrong, but I think it's correct to subtract the last line of the preamble from the line number of the error to get the offset post-preamble.